### ISSUE TITLE
Datasource Gitlab Project filters not working

### ISSUE THREAD

**Issue by phumberdroz** [gitlabhq/terraform-provider-gitlab#452]
Hi there,

### Terraform Version
```
Your version of Terraform is out of date! The latest version
is 0.13.4. You can update by downloading from https://www.terraform.io/downloads.html
Terraform v0.13.0
+ provider registry.terraform.io/gitlabhq/gitlab v3.0.0
+ provider registry.terraform.io/hashicorp/aws v2.70.0
+ provider registry.terraform.io/terraform-providers/gitlab v2.11.0
```

### Affected Resource(s)
Please list the resources as a list, for example:
- data.gitlab_projects

### Terraform Configuration Files
```hcl
data "gitlab_group" "company" {
  full_path = "company/services"
}

data "gitlab_projects" "company_projects" {
  group_id          = data.gitlab_group.company.id
  archived          = false
  with_shared = false
}
```

### Debug Output
Please provider a link to a GitHub Gist containing the complete debug output: https://www.terraform.io/docs/internals/debugging.html. Please do NOT paste the debug output in the issue; just paste a link to the Gist.
Included this:
```
Link: <https://gitlab.com/api/v4/groups/XXXXXX/projects?id=XXXXX&include_subgroups=false&order_by=created_at&owned=false&page=1&per_page=20&simple=false&sort=desc&starred=false&with_custom_attributes=false&with_issues_enabled=false&with_merge_requests_enabled=false&with_security_reports=false&with_shared=true>; rel="prev", <https://gitlab.com/api/v4/groups/XXXXXXX/projects?id=XXXXX&include_subgroups=false&order_by=created_at&owned=false&page=3&per_page=20&simple=false&sort=desc&starred=false&with_custom_attributes=false&with_issues_enabled=false&with_merge_requests_enabled=false&with_security_reports=false&with_shared=true>; rel="next", <https://gitlab.com/api/v4/groups/XXXXXX/projects?id=XXXXX&include_subgroups=false&order_by=created_at&owned=false&page=1&per_page=20&simple=false&sort=desc&starred=false&with_custom_attributes=false&with_issues_enabled=false&with_merge_requests_enabled=false&with_security_reports=false&with_shared=true>; rel="first", <https://gitlab.com/api/v4/groups/XXXXXX/projects?id=XXXXXXX&include_subgroups=false&order_by=created_at&owned=false&page=3&per_page=20&simple=false&sort=desc&starred=false&with_custom_attributes=false&with_issues_enabled=false&with_merge_requests_enabled=false&with_security_reports=false&with_shared=true>; rel="last"
```

### Expected Behavior
It should set the query param with_shared to false and should not include archived repos
### Actual Behavior

it included archived and shared projects

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1. `terraform apply`

**Comment by phumberdroz** [comment]
What I am doing for now is:

```hcl
locals {
  filter_1 = [for x in data.gitlab_projects.company_projects.projects: x if !x.archived]
  filter_2 = [for x in local.filter_1: x if length(regexall("company.*", x.path_with_namespace)) > 0 ]
  map = { for project in local.filter_2 : project.path_with_namespace => project }
}
```

**Comment by RyPeck** [comment]
I am experiencing the same issue.

Data Source

```
data "gitlab_projects" "group_projects" {
  group_id          = <group-id>
  archived          = false
  include_subgroups = true
  with_shared       = false
}
```

Trace Log

```
data.gitlab_projects.group_projects: Refreshing state...
2020-10-27T22:15:14.350-0400 [DEBUG] plugin.terraform-provider-gitlab_v3.1.0: 2020/10/27 22:15:14 [DEBUG] Reading Gitlab projects
2020-10-27T22:15:14.351-0400 [DEBUG] plugin.terraform-provider-gitlab_v3.1.0: 2020/10/27 22:15:14 [DEBUG] GitLab API Request Details:
2020-10-27T22:15:14.351-0400 [DEBUG] plugin.terraform-provider-gitlab_v3.1.0: ---[ REQUEST ]---------------------------------------
2020-10-27T22:15:14.351-0400 [DEBUG] plugin.terraform-provider-gitlab_v3.1.0: GET /api/v4/groups/<group_id>/projects?include_subgroups=true&page=1&per_page=100 HTTP/1.1
```

**Comment by RyPeck** [comment]
Switching the func `GetOk` to `GetOkExists` on the conditional [here](https://github.com/gitlabhq/terraform-provider-gitlab/blob/master/gitlab/data_source_gitlab_projects.go#L696) fixes the issue from my testing. But `GetOkExists` is deprecated according to the documentation. Any guidance on a possible fix?

cc @oboukili

**Comment by oboukili** [comment]
Brings back memories! I will work on that starting tomorrow, this datasource actually uses a non standard way to get arguments compared to other datasources/resources. Maybe just changing that will solve that issue. I'll try and add more acceptance tests as well.

**Comment by dawidmalina** [comment]
Any update on this issue? I am experiencing the same problem :(

**Comment by danw1979** [comment]
Still not resolved ?  tested datasource `gitlab_projects` in 3.8.0 yesterday with `archived = false` and still getting archived repos...

**Comment by timofurrer** [comment]
The problem here is that without `GetOkExists()` (which is deprecated) there isn't a way to check if a boolean attribute has been set to the default or set by the user to the same value as the default, which in our case is required, because the upstream API `archived` field actually has three states, too: `not set` (filter for all projects neglecting the `archived` state`), `true` (only filter for `archived` projects), `false` (only filter for `unarchived` projects).

There is a not really active issue here:

https://github.com/hashicorp/terraform-plugin-sdk/issues/817

So, maybe we'll get some answers ...

In the meantime what could be possible solutions is to have attributes like `only_archived`, `only_unarchived` fields. However, that clearly deviates from the provider best practices to use the same attributes in the provider as in the upstream APIs ...

**Comment by timofurrer** [comment]
... or we just make the attribute a `schema.TypeString` and one has to use `true` / `false` as strings?

**Comment by timofurrer** [comment]
It's seems like the `GetOkExists()` works for us and isn't going to go anywhere without a replacement, see https://github.com/hashicorp/terraform-plugin-sdk/pull/350#issuecomment-597888969

I've created a PR to fix it: https://github.com/gitlabhq/terraform-provider-gitlab/pull/855

**Comment by xM8WVqaG** [comment]
I know this is fairly typical, but GitLab just fixed the root cause of the issue. This API now returns even after a project has been archived.

Merged and released MR here: https://gitlab.com/gitlab-org/gitlab/-/merge_requests/80652

I tested setting `archived = true` on a gitlab.com hosted project with an older version of the GitLab provider and was able to archive, unarchive, and plan in between without issue.

Versions tested with:
```
Terraform v1.1.3
on linux_amd64
+ provider registry.terraform.io/gitlabhq/gitlab v3.8.0
```

**Comment by timofurrer** [comment]
@xM8WVqaG thanks for the information! That's good news. I agree that in this case the fix wouldn't be necessary for GitLab >= 14.8 and gitlab.com. However, we try support *some* of the latest GitLab releases, too - not only the latest and greatest :)
